File: sandboxes/node/service_runner.js
const fs = require('fs');
const http = require('http');
const { spawn } = require('child_process');
const net = require('net');

function getFreePort() {
  return new Promise((resolve) => {
    const srv = net.createServer();
    srv.listen(0, () => {
      const port = srv.address().port;
      srv.close(() => resolve(port));
    });
  });
}

function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function checkTcp(port) {
  return new Promise((resolve) => {
    const socket = new net.Socket();
    socket.setTimeout(100);
    socket.on('connect', () => {
      socket.destroy();
      resolve(true);
    });
    socket.on('timeout', () => {
      socket.destroy();
      resolve(false);
    });
    socket.on('error', () => {
      socket.destroy();
      resolve(false);
    });
    socket.connect(port, '127.0.0.1');
  });
}

function checkHttp(port, path) {
  return new Promise((resolve) => {
    const req = http.get({
      host: '127.0.0.1',
      port: port,
      path: path,
      timeout: 100
    }, (res) => {
      if (res.statusCode < 500) {
        resolve(true);
      } else {
        resolve(false);
      }
    });
    req.on('error', () => resolve(false));
    req.on('timeout', () => {
      req.abort();
      resolve(false);
    });
  });
}

function makeRequest(port, reqData) {
  return new Promise((resolve) => {
    const method = reqData.method || 'GET';
    const path = reqData.path || '/';
    const headers = reqData.headers || {};
    let bodyData = null;

    if (reqData.body) {
      if (typeof reqData.body === 'object') {
        bodyData = JSON.stringify(reqData.body);
        if (!headers['Content-Type']) headers['Content-Type'] = 'application/json';
      } else {
        bodyData = String(reqData.body);
      }
      headers['Content-Length'] = Buffer.byteLength(bodyData);
    }

    const options = {
      hostname: '127.0.0.1',
      port: port,
      path: path,
      method: method,
      headers: headers,
      timeout: 5000
    };

    const req = http.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        let parsedBody = data;
        try { parsedBody = JSON.parse(data); } catch (e) {}
        resolve({
          request: reqData,
          response: {
            status: res.statusCode,
            headers: res.headers,
            body: parsedBody
          }
        });
      });
    });

    req.on('error', (e) => {
      resolve({
        request: reqData,
        response: { status: 500, error: e.message }
      });
    });

    req.on('timeout', () => {
      req.abort();
      resolve({
        request: reqData,
        response: { status: 504, error: 'Timeout' }
      });
    });

    if (bodyData) {
      req.write(bodyData);
    }
    req.end();
  });
}

async function main() {
  const logs = [];
  const startTime = Date.now();
  let config;
  const workspaceDir = process.env.WORKSPACE_DIR || '.';

  try {
    const raw = fs.readFileSync(`${workspaceDir}/runtime.json`, 'utf8');
    config = JSON.parse(raw);
  } catch (e) {
    console.log(JSON.stringify({ status: 'error', logs: [`Failed to read runtime.json: ${e.message}`] }));
    process.exit(1);
  }

  const port = config.port || await getFreePort();
  const cmd = config.command;
  const health = config.health || {};
  const healthType = health.type || 'http';
  const healthPath = health.path || '/health';

  const env = Object.assign({}, process.env, { PORT: port });

  const dbProcesses = [];

  // 1. Redis
  if (config.redis) {
    logs.push("Starting in-memory Redis on port 6379...");
    env.REDIS_URL = "redis://127.0.0.1:6379";
    try {
      const rProc = spawn('redis-server', ['--port', '6379', '--save', ''], { stdio: 'ignore' });
      dbProcesses.push(rProc);
    } catch (e) {
      logs.push(`Failed to start Redis: ${e.message}`);
    }
  }

  // 2. SQLite
  const sqliteConfig = config.sqlite || {};
  const defaultSeedPath = path.join(workspaceDir, 'seed.sql');
  if (Object.keys(sqliteConfig).length > 0 || fs.existsSync(defaultSeedPath)) {
    const dbFile = sqliteConfig.db_file || 'db.sqlite';
    const seedFile = sqliteConfig.seed_file || 'seed.sql';
    const dbPath = `${workspaceDir}/${dbFile}`;
    
    env.DATABASE_URL = `sqlite:///${dbPath}`;

    if (fs.existsSync(dbPath)) {
      try { fs.unlinkSync(dbPath); } catch(e){}
    }

    if (seedFile) {
      const seedPath = `${workspaceDir}/${seedFile}`;
      logs.push(`Hydrating SQLite database from ${seedFile}...`);
      if (fs.existsSync(seedPath)) {
        try {
          const sqliteProc = spawn('sqlite3', [dbPath]);
          const seedContent = fs.readFileSync(seedPath);
          sqliteProc.stdin.write(seedContent);
          sqliteProc.stdin.end();
          // wait synchronously via a simple loop or just let it finish. 
          // For simplicity in a script, we can use spawnSync instead
          const { spawnSync } = require('child_process');
          const res = spawnSync('sqlite3', [dbPath], { input: seedContent });
          if (res.status !== 0) {
            logs.push(`Warning: SQLite seed failed: ${res.stderr.toString()}`);
          }
        } catch (e) {
          logs.push(`Warning: SQLite seed failed: ${e.message}`);
        }
      } else {
        logs.push(`Warning: SQLite seed file ${seedFile} not found.`);
      }
    }
  }

  logs.push(`Starting server with PORT=${port}...`);

  const serverProcess = spawn(cmd[0], cmd.slice(1), {
    env: env,
    cwd: workspaceDir,
    stdio: 'pipe'
  });

  let stdoutStr = '';
  let stderrStr = '';
  serverProcess.stdout.on('data', d => stdoutStr += d.toString());
  serverProcess.stderr.on('data', d => stderrStr += d.toString());

  let processExited = false;
  let exitCode = null;
  serverProcess.on('exit', (code) => {
    processExited = true;
    exitCode = code;
  });

  let isReady = false;
  for (let i = 0; i < 50; i++) {
    if (processExited) break;
    
    if (healthType === 'tcp') {
      isReady = await checkTcp(port);
    } else {
      isReady = await checkHttp(port, healthPath);
    }

    if (isReady) break;
    await delay(100);
  }

  const startupTime = Date.now() - startTime;

  if (!isReady) {
    serverProcess.kill();
    for (const p of dbProcesses) {
      try { p.kill(); } catch(e){}
    }
    logs.push(`Server failed to start or pass health check within 5 seconds.`);
    console.log(JSON.stringify({
      status: 'error',
      startupTime: startupTime,
      stdout: stdoutStr.split('\n'),
      stderr: stderrStr.split('\n'),
      logs: logs,
      exitCode: exitCode
    }));
    process.exit(1);
  }

  logs.push(`Server ready in ${startupTime}ms.`);

  let requests = [];
  try {
    const rawRequests = fs.readFileSync(`${workspaceDir}/stdin.txt`, 'utf8');
    if (rawRequests.trim()) {
      requests = JSON.parse(rawRequests);
    }
  } catch (e) {
    logs.push(`Warning: Failed to parse stdin.txt requests: ${e.message}`);
  }

  const responses = [];
  for (const req of requests) {
    responses.push(await makeRequest(port, req));
  }

  serverProcess.kill();
  
  // Wait slightly for exit
  for(let i=0; i<10; i++) {
    if (processExited) break;
    await delay(100);
  }

  for (const p of dbProcesses) {
    try { p.kill(); } catch(e){}
  }

  console.log(JSON.stringify({
    status: 'success',
    startupTime: startupTime,
    responses: responses,
    stdout: stdoutStr.split('\n').filter(Boolean),
    stderr: stderrStr.split('\n').filter(Boolean),
    logs: logs,
    exitCode: exitCode || 0
  }));
}

main().catch(e => {
  console.log(JSON.stringify({ status: 'error', logs: [e.message] }));
  process.exit(1);
});

File: sandboxes/python/service_runner.py
import sys
import json
import time
import urllib.request
import subprocess
import socket
import os

def get_free_port():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(("", 0))
        return s.getsockname()[1]

def main():
    logs = []
    start_time = time.time()
    
    workspace_dir = os.environ.get("WORKSPACE_DIR", ".")
    try:
        with open(f"{workspace_dir}/runtime.json") as f:
            config = json.load(f)
    except Exception as e:
        print(json.dumps({"status": "error", "logs": [f"Failed to read runtime.json: {e}"]}))
        sys.exit(1)

    port = config.get("port", get_free_port())
    cmd = config.get("command")
    health = config.get("health", {})
    health_path = health.get("path", "/health")
    health_type = health.get("type", "http")

    env = dict(os.environ)
    env["PORT"] = str(port)

    # ── Phase 4.1 Database Mocking ──
    db_processes = []
    
    # 1. Redis
    redis_config = config.get("redis", False)
    if redis_config:
        logs.append("Starting in-memory Redis on port 6379...")
        env["REDIS_URL"] = "redis://127.0.0.1:6379"
        try:
            r_proc = subprocess.Popen(
                ["redis-server", "--port", "6379", "--save", ""],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL
            )
            db_processes.append(r_proc)
        except Exception as e:
            logs.append(f"Failed to start Redis: {e}")

    # 2. SQLite
    sqlite_config = config.get("sqlite", {})
    if sqlite_config or os.path.exists(f"{workspace_dir}/seed.sql"):
        db_file = sqlite_config.get("db_file", "db.sqlite")
        seed_file = sqlite_config.get("seed_file", "seed.sql")
        
        db_path = f"{workspace_dir}/{db_file}"
        env["DATABASE_URL"] = f"sqlite:///{db_path}"
        
        if os.path.exists(db_path):
            os.remove(db_path)
            
        if seed_file:
            seed_path = f"{workspace_dir}/{seed_file}"
            logs.append(f"Hydrating SQLite database from {seed_file}...")
            if os.path.exists(seed_path):
                try:
                    with open(seed_path, "r") as sf:
                        subprocess.run(
                            ["sqlite3", db_path],
                            stdin=sf,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE,
                            check=True
                        )
                except subprocess.CalledProcessError as e:
                    logs.append(f"Warning: SQLite seed failed: {e.stderr.decode('utf-8', errors='ignore')}")
            else:
                logs.append(f"Warning: SQLite seed file {seed_file} not found.")

    # Write a simple shell wrapper to ensure command is found
    # If cmd is an array, we execute it directly
    logs.append(f"Starting server with PORT={port}...")
    try:
        process = subprocess.Popen(
            cmd,
            env=env,
            cwd=workspace_dir,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )
    except Exception as e:
        print(json.dumps({"status": "error", "logs": [f"Failed to start server: {e}"]}))
        sys.exit(1)

    # Health check polling
    is_ready = False
    for _ in range(50): # 5 seconds max
        try:
            if health_type == "tcp":
                with socket.create_connection(("127.0.0.1", port), timeout=1.0):
                    is_ready = True
                    break
            else:
                req = urllib.request.Request(f"http://127.0.0.1:{port}{health_path}", method="GET")
                with urllib.request.urlopen(req, timeout=1.0) as res:
                    if res.status < 500:
                        is_ready = True
                        break
        except Exception:
            pass
        
        # Check if process crashed
        if process.poll() is not None:
            break
            
        time.sleep(0.1)

    startup_time_ms = int((time.time() - start_time) * 1000)

    if not is_ready:
        process.terminate()
        stdout, stderr = process.communicate()
        logs.append(f"Server failed to start or pass health check within 5 seconds.")
        
        for p in db_processes:
            try:
                p.terminate()
            except:
                pass
                
        print(json.dumps({
            "status": "error",
            "startupTime": startup_time_ms,
            "stdout": stdout.splitlines() if stdout else [],
            "stderr": stderr.splitlines() if stderr else [],
            "logs": logs,
            "exitCode": process.returncode
        }))
        sys.exit(1)

    logs.append(f"Server ready in {startup_time_ms}ms.")

    # Read requests from stdin.txt
    requests = []
    try:
        with open(f"{workspace_dir}/stdin.txt") as f:
            raw = f.read()
            if raw.strip():
                requests = json.loads(raw)
    except Exception as e:
        logs.append(f"Warning: Failed to parse stdin.txt requests: {e}")

    # Execute requests
    responses = []
    for req in requests:
        method = req.get("method", "GET")
        path = req.get("path", "/")
        headers = req.get("headers", {})
        body = req.get("body", None)

        url = f"http://127.0.0.1:{port}{path}"
        
        data = None
        if body:
            if isinstance(body, dict) or isinstance(body, list):
                data = json.dumps(body).encode("utf-8")
                if "Content-Type" not in headers:
                    headers["Content-Type"] = "application/json"
            else:
                data = str(body).encode("utf-8")

        req_obj = urllib.request.Request(url, data=data, headers=headers, method=method)
        
        res_data = {
            "request": req,
            "response": {}
        }
        
        try:
            with urllib.request.urlopen(req_obj, timeout=5.0) as res:
                res_body = res.read().decode("utf-8")
                try:
                    parsed_body = json.loads(res_body)
                except:
                    parsed_body = res_body
                    
                res_data["response"] = {
                    "status": res.status,
                    "headers": dict(res.getheaders()),
                    "body": parsed_body
                }
        except urllib.error.HTTPError as e:
            res_body = e.read().decode("utf-8")
            try:
                parsed_body = json.loads(res_body)
            except:
                parsed_body = res_body
            res_data["response"] = {
                "status": e.code,
                "headers": dict(e.headers),
                "body": parsed_body
            }
        except Exception as e:
            res_data["response"] = {
                "status": 500,
                "error": str(e)
            }
            
        responses.append(res_data)

    # Clean up
    process.terminate()
    try:
        stdout, stderr = process.communicate(timeout=2.0)
    except subprocess.TimeoutExpired:
        process.kill()
        stdout, stderr = process.communicate()
        
    for p in db_processes:
        try:
            p.terminate()
            p.kill()
        except:
            pass

    # Return structured result
    print(json.dumps({
        "status": "success",
        "startupTime": startup_time_ms,
        "responses": responses,
        "stdout": stdout.splitlines() if stdout else [],
        "stderr": stderr.splitlines() if stderr else [],
        "logs": logs,
        "exitCode": process.returncode
    }))

if __name__ == "__main__":
    main()
